src/System.Slices/System/Buffers/Memory.cs

```diff
@@ -55,7 +55,7 @@ public Memory<T> Slice(int index, int length)
 
         public Span<T> Span => _owner.GetSpanInternal(_id).Slice(_index, _length);
 
-        public DisposableReservation Reserve() => new DisposableReservation(_owner, _id);
+        public DisposableReservation<T> Reserve() => new DisposableReservation<T>(_owner, _id);
```

Comment:
I used to have this type as a generic type and it was a problem. Somethimes you want to store many reservations on a collection, and generic reservations mean you need many different types of collections.